Emby.Server.Implementations/Library/Resolvers/AudioBooks/AudioBookResolver.cs

```diff
@@ -0,0 +1,145 @@
+    /// </summary>
+    public class AudioBookResolver : GenericFolderResolver<AudioBook>
+    {
+        private readonly ILogger<AudioBookResolver> _logger;
+        private readonly NamingOptions _namingOptions;
+        private readonly IDirectoryService _directoryService;
+
+        public AudioBookResolver(ILogger<AudioBookResolver> logger, NamingOptions namingOptions, IDirectoryService directoryService)
+        {
+            _logger = logger;
+            _namingOptions = namingOptions;
+            _directoryService = directoryService;
+        }
+
+        /// <summary>
+        /// Attempt to resolve a single audio book from args.Path.
+        /// Expected cases:
+        /// Path is an AudioBook directory: Resolve path as AudioBook and children as AudioBookFiles.
+        /// Path is single-file AudioBook: Resolve path as AudioBook w/ AudioBookFile of path.
+        /// </summary>
+        /// <param name="args">The args.</param>
+        /// <returns>Entities.AudioBook.</returns>
+        protected override AudioBook Resolve(ItemResolveArgs args)
+        {
+            // We only care about audiobooks
+            if (args.GetCollectionType() != CollectionType.books)
+            {
+                return null;
+            }
+
+            // TODO: I don't know if there's a valid case where parent is null and CollectionType is books
+            if (args.Parent != null && args.Parent.GetType() != typeof(MediaBrowser.Controller.Entities.Folder))
+            {
+                return null;
+            }
+
+            var audioBook = new AudioBook();
+
+            // TODO: Get title from path or from child file "Album" metadata
+            var bookFileName = Path.GetFileNameWithoutExtension(args.Path);
+            var name = MatchFor(bookFileName, "name");
+            name ??= bookFileName;
+
+            var year = MatchFor(bookFileName, "year");
+
+            audioBook.Name = name;
+            audioBook.ProductionYear = Convert.ToInt16(year, CultureInfo.InvariantCulture);
+
+            // Process directory to resolve single audio book
+            if (args.IsDirectory)
+            {
+                // Majority children must be audio files
+                int audioFiles = 0;
+                int nonAudioFiles = 0;
+                foreach (var child in args.FileSystemChildren)
+                {
+                    if (AudioFileParser.IsAudioFile(child.FullName, _namingOptions))
+                    {
+                        audioFiles += 1;
+                    }
+                    else
+                    {
+                        nonAudioFiles += 1;
+                    }
+                }
+
+                if (audioFiles == 0 || nonAudioFiles / audioFiles > 1)
+                {
+                    _logger.LogDebug("Less than half of the files in {0} were audio files, probably not an AudioBook directory", args.Path);
```

Comment:
Why is this check needed?

Reply:
By my understanding, resolvers are attempted until one succeeds. The priority value is high (low in sorted list) so it shouldn't come up often, but I didn't want to assume it would always only be resolving audiobooks. This catches cases where there is an unresolved directory which contains at least one file with an audio extension but mostly other files which I believe is unlikely for an audiobook. I imagine most cases will be a directory containing some number of audiobook files (chapters / parts) and maybe a text file or two with some arbitrary information.